vm/devices/storage/disk_layered/src/lib.rs

```diff
@@ -153,47 +160,79 @@ impl LayeredDisk {
     ///
     /// The layers must be ordered from top to bottom, with the top layer being
     /// the first in the list.
-    pub fn new(
+    pub async fn new(
         read_only: bool,
-        mut layers: Vec<LayerConfiguration>,
+        layers: Vec<LayerConfiguration>,
     ) -> Result<Self, InvalidLayeredDisk> {
         if layers.is_empty() {
             return Err(InvalidLayeredDisk::NoLayers);
         }
 
-        // Collect the common properties of the layers.
+        let mut attached_layers: Vec<LayerConfiguration<AttachedDiskLayer>> = {
+            let mut attached_layers = Vec::new();
+            // layers are attached to one another from the bottom-up, hence the need
+            // to iterate in reverse.
+            let mut lower_layer_metadata = None;
+            for (i, config) in layers.into_iter().enumerate().rev() {
+                let layer = config
+                    .layer
+                    .0
+                    .attach(lower_layer_metadata.take())
+                    .await
+                    .map_err(|e| InvalidLayeredDisk::Layer(i, InvalidLayer::AttachFailed(e)))?;
+                lower_layer_metadata = Some(layer.meta.clone());
```

Comment:
Should we do some checking on the metadata here rather than wait until we have all the metadata? This would allow the individual lazy layers to have some assurance that the input data had already been validated.

Reply:
Mainly I think it would be useful to move the sector size power of 2 check here.

Reply:
Yeah, I touched on this a bit in https://github.com/microsoft/openvmm/pull/423#discussion_r1871849310, but I wanted to get your thoughts.

Looking at the code, it seems reasonable to check / collect data while attaching disks? i.e: that validation need not be from the top-down?

Reply:
I think it's easier to reason about some of the validation top down, because we look at some things only up to and including the first non-write-through layer.

But sector size could certainly be validated bottom up.

Reply:
I.e., when you look at a layer in bottom up iteration, you don't know if we're going to end up writing to it.

Reply:
Sounds good. I'll move some of the validation into this loop.

Reply:
I moved all layer-specific validation into the bottom-up attach loop, leaving the top-down loop to collect disk info and perform layer-stack dependent validation.